### PR TITLE
Upgrade jsonlint and integrate its new features (json5, better errors)

### PR DESCRIPTION
**Pull request by prantlf**
This is the rest that I added in my fork. This should be the last PR in in the quick sequence :-)

* Upgrade `jsonlint` and `eslint` dependencies. Jsonlint contains a different parser with better performance, error handling and new features.
* Introduce a new option - `ignoreTrailingCommas`, false by default.
* Introduce a new option - `allowDuplicateObjectKeys`, true by default.
* Pass the original JSON string to the schema validator to enable reporting the location and context of the error occurrence.
* Include column number of the error occurrence in formatters.
* Use the structured error information for both data parsing and schema validation in the jshint reporter.
* Make the plain reporter return a string too - the exception message.
* Add unit tests for the new features and update unit tests for changes in error formatting.

### CODE REVIEW

**Comment by brandonramirez** [comment]
Sorry for the delay - been busy.  This will take a while to review, but I'll review it.  Thank you so much for contributing!

File: lib/reporters/jshint-style.js [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
   }
-  const line = grunt.file.read(file).split(/\r?\n/)[errorDetails.loc.first_line];  // intentionally do not read it as json.
-  return `     ${errorDetails.line} |     ${errorDetails.text}
-             ^ Expected ${grunt.log.wordlist(errorDetails.expected)} and instead saw '${line.charAt(errorDetails.loc.first_column + 1)}'`;
+  return value;
+}
+
+module.exports = function reportLikeJshint(file, error) {
+  let line;
+  let exzerpt;
```
Comment (brandonramirez) [inline lib/reporters/jshint-style.js]: This is the German spelling. - should be `excerpt` in English.

File: test/unit-tests.js [hunk 2]
```diff
@@ -160,35 +160,22 @@ function createPassingJsonlintSpy() {
...
-          last_column: 11
-        },
-        expected: [
-          "']'"
-        ]
+  return {
+    parse() {
+      const error = new SyntaxError('Parse error on line 1, column 3:\n{ 3...\n--^\nExpected "}" and instead saw "3"');
+      error.reason = 'Expected "}" and instead saw "3"';
+      error.exzerpt = '{ 3...';
```
Comment (brandonramirez) [inline test/unit-tests.js]: `excerpt`

File: lib/grunt-jsonlint-task.js [hunk 3]
```diff
@@ -26,70 +26,65 @@ module.exports = function makeTask(grunt, jsonlint, validator, sorter) {
...
+    if (this.filesSrc) {
       let failed = 0;
-      let errorDetails = null;
-
-      /* eslint no-param-reassign:
-        ["error", { "props": true, "ignorePropertyModificationsFor": ["jsonlint"] }]
-      */
-
-      // Monkey patch the parseError function to something we can control.
-      const originalParseError = jsonlint.parser.yy.parseError;
```
Comment (brandonramirez) [inline lib/grunt-jsonlint-task.js]: Excellent!  I'm glad you were able to get rid of this old approach for over-riding `parseErrror`.

File: lib/reporters/jshint-style.js [hunk 4]
```diff
@@ -1,8 +1,34 @@
...
+  let pointer;
+  let reason;
+  if (error.location) {
+    ({ line } = error.location.start);
```
Comment (brandonramirez) [inline lib/reporters/jshint-style.js]: Extra parenthesis?

File: lib/reporters/jshint-style.js [hunk 5]
```diff
@@ -1,8 +1,34 @@
...
+  if (extraLength > 0) {
+    pointer = `   ${pointer.substr(extraLength)}`;
+    exzerpt = `...${exzerpt.substr(extraLength)}`;
+  }
+  if (exzerpt.length > 20) {
+    exzerpt = `${exzerpt.substr(0, 20)}...`;
+  }
+  pointer = pointer.replace(/-/g, ' ');
+  const prefix = `${padLeft(line, 6)} | `;
+  const indent = ' '.repeat(prefix.length);
```
Comment (brandonramirez) [inline lib/reporters/jshint-style.js]: TIL about `String.prototype.repeat`!

File: package.json [hunk 6]
```diff
@@ -7,8 +7,12 @@
     "test": "nyc grunt test",
     "fix-lint-errors": "eslint --fix tasks/ lib/ test/*.js"
   },
+  "files": [
```
Comment (brandonramirez) [inline package.json]: Cool!  I didn't know about the package.json `files` block.  I definitely prefer it to `.npmignore`.

File: README.md [hunk 7]
```diff
@@ -67,16 +66,49 @@ Add the following (multi-)task to your `Gruntfile`:
...
           indent: 2,
-          sortKeys: false
+          sortKeys: false,
+          pruneComments: false,
+          stripObjectKeys: false
         }
       }
     }
 
+<<<<<<< HEAD
```
Comment (brandonramirez) [inline README.md]: Merge markers - incomplete merge?

File: README.md [hunk 8]
```diff
@@ -83,32 +76,12 @@ Add the following (multi-)task to your `Gruntfile`:
       }
     }
 
-<<<<<<< HEAD
```
Comment (brandonramirez) [inline README.md]: Great - you cleaned up the merge problems.  Thank you.

File: package.json (hunk 6 above)
Comment (prantlf) [reply package.json]: It's preferable, isn't it? I see whitelisting less-error-prone to blacklisting too.

File: README.md (hunk 8 above)
Comment (prantlf) [reply README.md]: Oops, sorry, I must've missed it. Git chould detect it itself...

File: README.md (hunk 7 above)
Comment (prantlf) [reply README.md]: Oops, sorry, I must've missed it. Git could detect it itself...

**Pull request merged**